### PR TITLE
DRILL-4618: Fix hive function loader not correctly take random flag; …

### PR DESCRIPTION
**Pull request by chunhui-shi**
…and function visitor should not use previous function holder if this function is nondeterministic

### CODE REVIEW

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/EvaluationVisitor.java [hunk 1]
```diff
@@ -811,7 +811,7 @@ public HoldingContainer visitFunctionCall(FunctionCall call, ClassGenerator<?> g
     @Override
     public HoldingContainer visitFunctionHolderExpression(FunctionHolderExpression holder, ClassGenerator<?> generator) throws RuntimeException {
       HoldingContainer hc = getPrevious(holder, generator.getMappingSet());
-      if (hc == null) {
+      if (hc == null || holder.isRandom()) {
```
Comment (StevenMPhillips) [inline exec/java-exec/src/main/java/org/apache/drill/exec/expr/EvaluationVisitor.java]: I think this change is not sufficient to cover all cases. If random() is just part of the expression, then this won't help. For example, the expression 2 * random().

I think a better way to fix this would be to modify EqualityVisitor.visitFunctionHolderExpression() to return false if the function is random. That way any expression which contains a random function will never be considered equal to another expression.

**Comment by chunhui-shi** [comment]
@StevenMPhillips Hi Steven, is this fix addressing your concern now? Thanks.

**Comment by Ben-Zvi** [comment]
Tested and checked ... LGTM

File: contrib/storage-hive/core/src/test/java/org/apache/drill/exec/fn/hive/TestInbuiltHiveUDFs.java [hunk 2]
```diff
@@ -84,7 +84,6 @@ public void testGetJsonObject() throws Exception {
         .go();
   }
 
-  @Test // DRILL-3272
   public void testIf() throws Exception {
```
Comment (sudheeshkatkam) [inline contrib/storage-hive/core/src/test/java/org/apache/drill/exec/fn/hive/TestInbuiltHiveUDFs.java]: Why is this test being ignored? Is this test not relevant anymore?

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/EqualityVisitor.java [hunk 3]
```diff
@@ -75,6 +75,9 @@ public Boolean visitFunctionHolderExpression(FunctionHolderExpression holder, Lo
     if (!holder.getName().equals(((FunctionHolderExpression) value).getName())) {
       return false;
     }
+    if(holder.isRandom()) {
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/expr/EqualityVisitor.java]: spacing: `if (`

Comment (chunhui-shi) [inline exec/java-exec/src/main/java/org/apache/drill/exec/expr/EqualityVisitor.java]: Fixed.

File: contrib/storage-hive/core/src/test/java/org/apache/drill/exec/fn/hive/TestInbuiltHiveUDFs.java (hunk 2 above)
Comment (chunhui-shi) [inline contrib/storage-hive/core/src/test/java/org/apache/drill/exec/fn/hive/TestInbuiltHiveUDFs.java]: It was a merge error. Fixed.

File: contrib/storage-hive/core/src/test/java/org/apache/drill/exec/fn/hive/TestInbuiltHiveUDFs.java [hunk 4]
```diff
@@ -84,7 +84,7 @@ public void testGetJsonObject() throws Exception {
         .go();
   }
 
-  @Test // DRILL-3272
+  @Test //DRILL-4618
```
Comment (sudheeshkatkam) [inline contrib/storage-hive/core/src/test/java/org/apache/drill/exec/fn/hive/TestInbuiltHiveUDFs.java]: How is this test related to 4618?

**Comment by sudheeshkatkam** [comment]
+1
